gitbootcamp.rst

```diff
@@ -0,0 +1,268 @@
+
+
+Creating a Pull Request
+-----------------------
+
+1. Go to https://github.com/python/cpython.
+
+2. Click ``compare across forks`` link.
+
+3. Select the base fork: ``python/cpython`` and base branch: ``master``
+
+4. Select the head fork: ``<username>/cpython`` and base branch: the branch
+   containing your changes.
+
+5. Press ``Create Pull Request`` button.
+
+
+Syncing With Upstream
+---------------------
+
+Scenario:
+
+- You forked cpython repository some time ago.
+- Time passes.
+- There have been new commits made in upstream cpython repository.
+- Your forked cpython repository is no longer up to date.
+- You now want to update your forked cpython repository to be the same as
+  upstream.
+
+Solution::
+
+   $ git checkout master
+   $ git pull --rebase upstream master
+   $ git push origin master
+
+The `--rebase` is only needed if you have local changes to the branch.
+
+Another scenario:
+
+- You created ``some-branch`` some time ago.
+- Time passes.
+- You made some commits to ``some-branch``.
+- Meanwhile, there are recent changes from upstream cpython repository.
+- You want to incorporate the recent changes from upstream into ``some-branch``.
+
+Solution::
+
+   $ git checkout some-branch
+   $ git fetch upstream
+   $ git rebase upstream/master
+
+
+Backporting Merged Changes
+--------------------------
+
+When a pull request has been merged to master, and it needs to be backported
+into one of the maintenance branches.
+
+First, obtain the commit sha1 from the merged pull request:
+
+1. Go to the merged pull request page, for example::
+
+     https://github.com/python/cpython/pull/PR-ID
+
+2. Scroll down and find the activity that says something like::
+
+     CoreDeveloper merged commit <hash> into python:master ...
+
+3. Follow the link to <hash>.
+
+4. Copy the complete hash value.
+
+The commit hash will be used below.
+
+To backport the commit to 3.6::
+
+   $ git fetch upstream
+   $ git checkout -b backport-someissue-3.6 upstream/3.6
+   $ git cherry-pick -x hashvalue
+   $ git push origin backport-someissue-3.6
+
+Go to https://github.com/python/cpython to create the pull request.  Select
+``3.6`` as the base branch, and ``backport-someissue-3.6`` as the head branch.
+
+
+Downloading Other's Patches
+---------------------------
+
+Scenario:
+
+- A contributor made a pull request to cpython.
+- Before merging it, you want to be able to test their changes locally.
+
+Set up the following git alias::
+
+   $ git config --global alias.pr '!sh -c "git fetch upstream pull/${1}/head:pr_${1} && git checkout pr_${1}" -'
+
+The alias only needs to be done once.  After the alias is set up, you can get a
+local copy of a pull request as follows::
+
+   $ git pr <pr_number>
+
+For example, to fetch and checkout pull request #777::
+
+   $ git pr 777
+
+
+Accepting and Merging A Pull Request
+------------------------------------
+
+Pull requests can be accepted and merged by a Python Core Developer.
```

Comment:
A question about "merging etiquette": if another core-dev submitted a PR, and it looks good to me, should I go ahead and merge it, or should I just leave a positive review and let the original core-dev merge it?
I think the latter is better, but I've seen the former happen.
Perhaps this could be clarified here.